File: typescript/src/rules/definitions.ts
import type { Definition } from "./types.js";

export const ruleIDs = {
  readmeRequired: "repo.readme-required",
  agentsRequired: "repo.agents-required",
  ciRequired: "repo.ci-required",
  slophammerCiRequired: "repo.slophammer-ci-required",
  tsPackageRequired: "ts.package-required",
  tsTypecheckRequired: "ts.typecheck-required",
  tsStrictRequired: "ts.strict-required",
  tsNoExplicitAny: "ts.no-explicit-any",
  tsNoUnsafeTypes: "ts.no-unsafe-types",
  tsLintRequired: "ts.lint-required",
  tsFormatRequired: "ts.format-required",
  tsTestRequired: "ts.test-required",
  tsCoverageRequired: "ts.coverage-required",
  tsComplexityRequired: "ts.complexity-required",
  tsDryRequired: "ts.dry-required",
  tsMutationRequired: "ts.mutation-required",
  tsDependencyBoundariesRequired: "ts.dependency-boundaries-required",
  tsScopeIncomplete: "ts.scope-incomplete",
  tsSuppressionsJustified: "ts.suppressions-justified"
} as const;

export const defaultDefinitions: readonly Definition[] = [
  {
    id: ruleIDs.readmeRequired,
    title: "README required",
    category: "repo",
    severity: "error",
    path: "README.md",
    message: "README.md is required",
    description: "The target repo should have a README.md.",
    status: "implemented"
  },
  {
    id: ruleIDs.agentsRequired,
    title: "Agent instructions required",
    category: "repo",
    severity: "error",
    path: "AGENTS.md",
    message: "AGENTS.md is required",
    description: "The target repo should have an AGENTS.md.",
    status: "implemented"
  },
  {
    id: ruleIDs.ciRequired,
    title: "CI workflow required",
    category: "repo",
    severity: "error",
    path: ".github/workflows",
    message: ".github/workflows must contain at least one .yml or .yaml workflow",
    description: "The target repo should have a CI workflow under .github/workflows.",
    status: "implemented"
  },
  {
    id: ruleIDs.slophammerCiRequired,
    title: "Slophammer enforcement required",
    category: "repo",
    severity: "error",
    path: ".github/workflows",
    message: "CI must run a Slophammer checker when slophammer.yml is present",
    description:
      "A repository that carries slophammer.yml must execute a Slophammer checker from binding CI evidence; config without enforcement is decoration.",
    status: "implemented"
  },
  {
    id: ruleIDs.tsPackageRequired,
    title: "TypeScript package required",
    category: "typescript",
    severity: "error",
    path: "package.json",
    message: "TypeScript projects must include package.json",
    description: "TypeScript projects should include package.json.",
    status: "implemented"
  },
  {
    id: ruleIDs.tsTypecheckRequired,
    title: "TypeScript typecheck required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare a no-emit typecheck in CI or scripts",
    description:
      "TypeScript projects should declare tsc, tsgo, or an equivalent no-emit typecheck.",
    tool: "tsc/tsgo --noEmit",
    status: "implemented"
  },
  {
    id: ruleIDs.tsStrictRequired,
    title: "TypeScript strict mode required",
    category: "typescript",
    severity: "error",
    path: "tsconfig.json",
    message: "TypeScript projects must enable strict mode",
    description: "TypeScript projects should enable strict compiler settings.",
    tool: "tsc",
    status: "implemented"
  },
  {
    id: ruleIDs.tsNoExplicitAny,
    title: "No explicit any",
    category: "typescript",
    severity: "error",
    path: "eslint.config.mjs",
    message: "TypeScript projects must reject explicit any",
    description: "TypeScript projects should configure lint rules to reject explicit any.",
    tool: "eslint/oxlint",
    status: "implemented"
  },
  {
    id: ruleIDs.tsNoUnsafeTypes,
    title: "No unsafe type operations",
    category: "typescript",
    severity: "error",
    path: "eslint.config.mjs",
    message: "TypeScript projects must reject unsafe type operations",
    description:
      "TypeScript projects should configure lint rules to reject unsafe assignments, calls, member access, and returns.",
    tool: "eslint/oxlint",
    status: "implemented"
  },
  {
    id: ruleIDs.tsLintRequired,
    title: "TypeScript lint required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare a linter in CI or scripts",
    description:
      "TypeScript projects should declare ESLint, Oxlint, Biome, or an equivalent linter.",
    tool: "eslint/oxlint/biome",
    status: "implemented"
  },
  {
    id: ruleIDs.tsFormatRequired,
    title: "TypeScript formatter required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare a formatter check",
    description:
      "TypeScript projects should declare a formatter check through Prettier, Oxfmt, Biome, Dprint, or an equivalent tool.",
    tool: "prettier/oxfmt/biome/dprint",
    status: "implemented"
  },
  {
    id: ruleIDs.tsTestRequired,
    title: "TypeScript tests required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare tests in CI or scripts",
    description: "TypeScript projects should declare a real test runner command.",
    tool: "node --test/vitest/jest",
    status: "implemented"
  },
  {
    id: ruleIDs.tsCoverageRequired,
    title: "TypeScript coverage gate required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare a coverage gate",
    description: "TypeScript projects should declare a coverage gate with a target of at least 85.",
    tool: "c8/nyc/vitest/jest coverage",
    status: "implemented"
  },
  {
    id: ruleIDs.tsComplexityRequired,
    title: "TypeScript complexity required",
    category: "typescript",
    severity: "error",
    path: "eslint.config.mjs",
    message: "TypeScript projects must enforce complexity limits",
    description:
      "TypeScript projects should enforce complexity limits through a configured linter.",
    tool: "eslint/oxlint/biome",
    status: "implemented"
  },
  {
    id: ruleIDs.tsDryRequired,
    title: "TypeScript DRY check required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare a DRY check",
    description:
      "TypeScript projects should declare Slophammer's native copied-block duplicate detector.",
    tool: "slophammer-ts dry",
    status: "implemented"
  },
  {
    id: ruleIDs.tsMutationRequired,
    title: "TypeScript mutation check required",
    category: "typescript",
    severity: "error",
    path: ".github/workflows",
    message: "TypeScript projects must declare mutation testing",
    description:
      "TypeScript projects should declare TypeScript mutation testing, normally through StrykerJS. Only executing invocations count: list, scan, and dry-run forms cannot fail on a surviving mutant and are not evidence.",
    tool: "stryker",
    status: "implemented"
  },
  {
    id: ruleIDs.tsDependencyBoundariesRequired,
    title: "TypeScript dependency boundaries required",
    category: "typescript",
    severity: "error",
    path: "slophammer.yml",
    message: "TypeScript projects must respect configured dependency boundaries",
    description:
      "TypeScript projects should declare dependency boundaries in slophammer.yml and keep imports inside them.",
    status: "implemented"
  },
  {
    id: ruleIDs.tsScopeIncomplete,
    title: "TypeScript scope completeness",
    category: "typescript",
    severity: "error",
    path: "slophammer.yml",
    message:
      "Configured TypeScript scope must cover all production files or exclude them with reasons",
    description:
      "Every production TypeScript file must be in configured scope or covered by a conventional or reasoned exclude, so findings cannot be hidden by narrowing scope.",
    status: "implemented"
  },
  {
    id: ruleIDs.tsSuppressionsJustified,
    title: "TypeScript suppressions justified",
    category: "typescript",
    severity: "error",
    path: ".",
    message: "lint and type suppressions in production TypeScript code must carry a description",
    description:
      "eslint-disable, @ts-ignore, @ts-expect-error, biome-ignore, and oxlint-disable directives in production scope must carry a description; bare suppressions accumulate silently.",
    status: "implemented"
  }
];

File: rust/crates/slophammer-cli/src/rust_rules/definitions.rs
use crate::core::{RuleDefinition, Severity};

pub mod rule_ids {
    pub const README_REQUIRED: &str = "repo.readme-required";
    pub const AGENTS_REQUIRED: &str = "repo.agents-required";
    pub const CI_REQUIRED: &str = "repo.ci-required";
    pub const SLOPHAMMER_CI_REQUIRED: &str = "repo.slophammer-ci-required";
    pub const RUST_MANIFEST_REQUIRED: &str = "rust.manifest-required";
    pub const RUST_MSRV_REQUIRED: &str = "rust.msrv-required";
    pub const RUST_CHECK_REQUIRED: &str = "rust.check-required";
    pub const RUST_FMT_REQUIRED: &str = "rust.fmt-required";
    pub const RUST_CLIPPY_REQUIRED: &str = "rust.clippy-required";
    pub const RUST_TEST_REQUIRED: &str = "rust.test-required";
    pub const RUST_COVERAGE_REQUIRED: &str = "rust.coverage-required";
    pub const RUST_COMPLEXITY_REQUIRED: &str = "rust.complexity-required";
    pub const RUST_DRY_REQUIRED: &str = "rust.dry-required";
    pub const RUST_MUTATION_REQUIRED: &str = "rust.mutation-required";
    pub const RUST_UNSAFE_POLICY_REQUIRED: &str = "rust.unsafe-policy-required";
    pub const RUST_DEPENDENCY_AUDIT_REQUIRED: &str = "rust.dependency-audit-required";
    pub const RUST_DEPENDENCY_BOUNDARIES_REQUIRED: &str = "rust.dependency-boundaries-required";
    pub const RUST_SCOPE_INCOMPLETE: &str = "rust.scope-incomplete";
    pub const RUST_SUPPRESSIONS_JUSTIFIED: &str = "rust.suppressions-justified";
}

pub fn default_definitions() -> Vec<RuleDefinition> {
    vec![
        RuleDefinition {
            id: rule_ids::README_REQUIRED,
            title: "README required",
            category: "repo",
            severity: Severity::Error,
            path: "README.md",
            message: "README.md is required",
            description: "The target repo should have a README.md.",
            tool: None,
            status: "implemented",
        },
        RuleDefinition {
            id: rule_ids::AGENTS_REQUIRED,
            title: "Agent instructions required",
            category: "repo",
            severity: Severity::Error,
            path: "AGENTS.md",
            message: "AGENTS.md is required",
            description: "The target repo should have an AGENTS.md.",
            tool: None,
            status: "implemented",
        },
        RuleDefinition {
            id: rule_ids::CI_REQUIRED,
            title: "CI workflow required",
            category: "repo",
            severity: Severity::Error,
            path: ".github/workflows",
            message: ".github/workflows must contain at least one .yml or .yaml workflow",
            description: "The target repo should have a CI workflow under .github/workflows.",
            tool: None,
            status: "implemented",
        },
        RuleDefinition {
            id: rule_ids::SLOPHAMMER_CI_REQUIRED,
            title: "Slophammer enforcement required",
            category: "repo",
            severity: Severity::Error,
            path: ".github/workflows",
            message: "CI must run a Slophammer checker when slophammer.yml is present",
            description: "A repository that carries slophammer.yml must execute a Slophammer checker from binding CI evidence; config without enforcement is decoration.",
            tool: None,
            status: "implemented",
        },
        rust_definition(
            rule_ids::RUST_MANIFEST_REQUIRED,
            "Rust manifest required",
            "Cargo.toml",
            "Rust projects must include Cargo.toml",
            "Rust projects should include a Cargo manifest.",
            Some("cargo"),
        ),
        rust_definition(
            rule_ids::RUST_MSRV_REQUIRED,
            "Rust MSRV required",
            "Cargo.toml",
            "Rust projects must declare a minimum supported Rust version",
            "Rust projects should declare a minimum supported Rust version.",
            Some("cargo"),
        ),
        rust_definition(
            rule_ids::RUST_CHECK_REQUIRED,
            "Rust cargo check required",
            ".github/workflows",
            "Rust projects must declare cargo check in CI or scripts",
            "Rust projects should declare cargo check in an inspectable workflow or script.",
            Some("cargo check"),
        ),
        rust_definition(
            rule_ids::RUST_FMT_REQUIRED,
            "Rust formatter required",
            ".github/workflows",
            "Rust projects must declare cargo fmt --check in CI or scripts",
            "Rust projects should declare cargo fmt --check in CI or scripts.",
            Some("cargo fmt"),
        ),
        rust_definition(
            rule_ids::RUST_CLIPPY_REQUIRED,
            "Rust Clippy required",
            ".github/workflows",
            "Rust projects must declare cargo clippy in CI or scripts",
            "Rust projects should declare cargo clippy with warnings denied.",
            Some("cargo clippy"),
        ),
        rust_definition(
            rule_ids::RUST_TEST_REQUIRED,
            "Rust tests required",
            ".github/workflows",
            "Rust projects must declare cargo test in CI or scripts",
            "Rust projects should declare cargo test in CI or scripts.",
            Some("cargo test"),
        ),
        rust_definition(
            rule_ids::RUST_COVERAGE_REQUIRED,
            "Rust coverage gate required",
            ".github/workflows",
            "Rust projects must declare a coverage gate",
            "Rust projects should declare an enforceable coverage gate.",
            Some("cargo llvm-cov"),
        ),
        rust_definition(
            rule_ids::RUST_COMPLEXITY_REQUIRED,
            "Rust complexity required",
            ".github/workflows",
            "Rust projects must enforce complexity limits",
            "Rust projects should enforce complexity limits through Clippy or Slophammer policy.",
            Some("clippy"),
        ),
        rust_definition(
            rule_ids::RUST_DRY_REQUIRED,
            "Rust DRY check required",
            ".github/workflows",
            "Rust projects must declare a DRY check",
            "Rust projects should declare Slophammer's native Rust DRY check.",
            Some("slophammer-rs dry"),
        ),
        rust_definition(
            rule_ids::RUST_MUTATION_REQUIRED,
            "Rust mutation check required",
            ".github/workflows",
            "Rust projects must declare mutation testing",
            "Rust projects should declare mutation testing, normally through cargo-mutants.",
            Some("cargo mutants"),
        ),
        rust_definition(
            rule_ids::RUST_UNSAFE_POLICY_REQUIRED,
            "Rust unsafe policy required",
            "slophammer.yml",
            "Rust projects must declare and respect an unsafe-code policy",
            "Rust projects should declare and respect an unsafe-code policy in slophammer.yml.",
            None,
        ),
        rust_definition(
            rule_ids::RUST_DEPENDENCY_AUDIT_REQUIRED,
            "Rust dependency audit required",
            ".github/workflows",
            "Rust projects must declare dependency audit checks",
            "Rust projects should declare cargo audit or cargo deny.",
            Some("cargo audit/cargo deny"),
        ),
        rust_definition(
            rule_ids::RUST_DEPENDENCY_BOUNDARIES_REQUIRED,
            "Rust dependency boundaries required",
            "slophammer.yml",
            "Rust projects must respect configured dependency boundaries",
            "Rust projects should declare dependency boundaries in slophammer.yml and keep local dependencies inside them.",
            None,
        ),
        rust_definition(
            rule_ids::RUST_SCOPE_INCOMPLETE,
            "Rust scope completeness",
            "slophammer.yml",
            "Configured Rust scope must cover all production files or exclude them with reasons",
            "Every production Rust file must be in configured scope or covered by a conventional or reasoned exclude, so findings cannot be hidden by narrowing scope.",
            None,
        ),
        rust_definition(
            rule_ids::RUST_SUPPRESSIONS_JUSTIFIED,
            "Rust suppressions justified",
            ".",
            "allow attributes in production Rust code must carry a reason",
            "An #[allow(...)] attribute in production scope must carry an adjacent // reason comment or use #[expect(..., reason = \"...\")]; bare suppressions accumulate silently.",
            None,
        ),
    ]
}

pub fn definition(rule_id: &str) -> &'static RuleDefinition {
    match rule_id {
        rule_ids::README_REQUIRED => &README_REQUIRED,
        rule_ids::AGENTS_REQUIRED => &AGENTS_REQUIRED,
        rule_ids::CI_REQUIRED => &CI_REQUIRED,
        rule_ids::RUST_MANIFEST_REQUIRED => &RUST_MANIFEST_REQUIRED,
        rule_ids::RUST_MSRV_REQUIRED => &RUST_MSRV_REQUIRED,
        rule_ids::RUST_CHECK_REQUIRED => &RUST_CHECK_REQUIRED,
        rule_ids::RUST_FMT_REQUIRED => &RUST_FMT_REQUIRED,
        rule_ids::RUST_CLIPPY_REQUIRED => &RUST_CLIPPY_REQUIRED,
        rule_ids::RUST_TEST_REQUIRED => &RUST_TEST_REQUIRED,
        rule_ids::RUST_COVERAGE_REQUIRED => &RUST_COVERAGE_REQUIRED,
        rule_ids::RUST_COMPLEXITY_REQUIRED => &RUST_COMPLEXITY_REQUIRED,
        rule_ids::RUST_DRY_REQUIRED => &RUST_DRY_REQUIRED,
        rule_ids::RUST_MUTATION_REQUIRED => &RUST_MUTATION_REQUIRED,
        rule_ids::RUST_UNSAFE_POLICY_REQUIRED => &RUST_UNSAFE_POLICY_REQUIRED,
        rule_ids::RUST_DEPENDENCY_AUDIT_REQUIRED => &RUST_DEPENDENCY_AUDIT_REQUIRED,
        rule_ids::RUST_DEPENDENCY_BOUNDARIES_REQUIRED => &RUST_DEPENDENCY_BOUNDARIES_REQUIRED,
        rule_ids::SLOPHAMMER_CI_REQUIRED => &SLOPHAMMER_CI_REQUIRED,
        rule_ids::RUST_SCOPE_INCOMPLETE => &RUST_SCOPE_INCOMPLETE,
        rule_ids::RUST_SUPPRESSIONS_JUSTIFIED => &RUST_SUPPRESSIONS_JUSTIFIED,
        _ => &README_REQUIRED,
    }
}

fn rust_definition(
    id: &'static str,
    title: &'static str,
    path: &'static str,
    message: &'static str,
    description: &'static str,
    tool: Option<&'static str>,
) -> RuleDefinition {
    RuleDefinition {
        id,
        title,
        category: "rust",
        severity: Severity::Error,
        path,
        message,
        description,
        tool,
        status: "implemented",
    }
}

static README_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::README_REQUIRED,
    title: "README required",
    category: "repo",
    severity: Severity::Error,
    path: "README.md",
    message: "README.md is required",
    description: "The target repo should have a README.md.",
    tool: None,
    status: "implemented",
};
static AGENTS_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::AGENTS_REQUIRED,
    title: "Agent instructions required",
    category: "repo",
    severity: Severity::Error,
    path: "AGENTS.md",
    message: "AGENTS.md is required",
    description: "The target repo should have an AGENTS.md.",
    tool: None,
    status: "implemented",
};
static CI_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::CI_REQUIRED,
    title: "CI workflow required",
    category: "repo",
    severity: Severity::Error,
    path: ".github/workflows",
    message: ".github/workflows must contain at least one .yml or .yaml workflow",
    description: "The target repo should have a CI workflow under .github/workflows.",
    tool: None,
    status: "implemented",
};
static RUST_MANIFEST_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_MANIFEST_REQUIRED,
    title: "Rust manifest required",
    category: "rust",
    severity: Severity::Error,
    path: "Cargo.toml",
    message: "Rust projects must include Cargo.toml",
    description: "Rust projects should include a Cargo manifest.",
    tool: Some("cargo"),
    status: "implemented",
};
static RUST_MSRV_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_MSRV_REQUIRED,
    title: "Rust MSRV required",
    category: "rust",
    severity: Severity::Error,
    path: "Cargo.toml",
    message: "Rust projects must declare a minimum supported Rust version",
    description: "Rust projects should declare a minimum supported Rust version.",
    tool: Some("cargo"),
    status: "implemented",
};
static RUST_CHECK_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_CHECK_REQUIRED,
    title: "Rust cargo check required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare cargo check in CI or scripts",
    description: "Rust projects should declare cargo check in an inspectable workflow or script.",
    tool: Some("cargo check"),
    status: "implemented",
};
static RUST_FMT_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_FMT_REQUIRED,
    title: "Rust formatter required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare cargo fmt --check in CI or scripts",
    description: "Rust projects should declare cargo fmt --check in CI or scripts.",
    tool: Some("cargo fmt"),
    status: "implemented",
};
static RUST_CLIPPY_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_CLIPPY_REQUIRED,
    title: "Rust Clippy required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare cargo clippy in CI or scripts",
    description: "Rust projects should declare cargo clippy with warnings denied.",
    tool: Some("cargo clippy"),
    status: "implemented",
};
static RUST_TEST_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_TEST_REQUIRED,
    title: "Rust tests required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare cargo test in CI or scripts",
    description: "Rust projects should declare cargo test in CI or scripts.",
    tool: Some("cargo test"),
    status: "implemented",
};
static RUST_COVERAGE_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_COVERAGE_REQUIRED,
    title: "Rust coverage gate required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare a coverage gate",
    description: "Rust projects should declare an enforceable coverage gate.",
    tool: Some("cargo llvm-cov"),
    status: "implemented",
};
static RUST_COMPLEXITY_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_COMPLEXITY_REQUIRED,
    title: "Rust complexity required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must enforce complexity limits",
    description: "Rust projects should enforce complexity limits through Clippy or Slophammer policy.",
    tool: Some("clippy"),
    status: "implemented",
};
static RUST_DRY_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_DRY_REQUIRED,
    title: "Rust DRY check required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare a DRY check",
    description: "Rust projects should declare Slophammer's native Rust DRY check.",
    tool: Some("slophammer-rs dry"),
    status: "implemented",
};
static RUST_MUTATION_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_MUTATION_REQUIRED,
    title: "Rust mutation check required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare mutation testing",
    description: "Rust projects should declare mutation testing, normally through cargo-mutants. Only executing invocations count: list, scan, check, dry-run, and manifest-only forms cannot fail on a surviving mutant and are not evidence.",
    tool: Some("cargo mutants"),
    status: "implemented",
};
static RUST_UNSAFE_POLICY_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_UNSAFE_POLICY_REQUIRED,
    title: "Rust unsafe policy required",
    category: "rust",
    severity: Severity::Error,
    path: "slophammer.yml",
    message: "Rust projects must declare and respect an unsafe-code policy",
    description: "Rust projects should declare and respect an unsafe-code policy in slophammer.yml.",
    tool: None,
    status: "implemented",
};
static RUST_DEPENDENCY_AUDIT_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_DEPENDENCY_AUDIT_REQUIRED,
    title: "Rust dependency audit required",
    category: "rust",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "Rust projects must declare dependency audit checks",
    description: "Rust projects should declare cargo audit or cargo deny.",
    tool: Some("cargo audit/cargo deny"),
    status: "implemented",
};
static RUST_DEPENDENCY_BOUNDARIES_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_DEPENDENCY_BOUNDARIES_REQUIRED,
    title: "Rust dependency boundaries required",
    category: "rust",
    severity: Severity::Error,
    path: "slophammer.yml",
    message: "Rust projects must respect configured dependency boundaries",
    description: "Rust projects should declare dependency boundaries in slophammer.yml and keep local dependencies inside them.",
    tool: None,
    status: "implemented",
};
static SLOPHAMMER_CI_REQUIRED: RuleDefinition = RuleDefinition {
    id: rule_ids::SLOPHAMMER_CI_REQUIRED,
    title: "Slophammer enforcement required",
    category: "repo",
    severity: Severity::Error,
    path: ".github/workflows",
    message: "CI must run a Slophammer checker when slophammer.yml is present",
    description: "A repository that carries slophammer.yml must execute a Slophammer checker from binding CI evidence; config without enforcement is decoration.",
    tool: None,
    status: "implemented",
};
static RUST_SCOPE_INCOMPLETE: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_SCOPE_INCOMPLETE,
    title: "Rust scope completeness",
    category: "rust",
    severity: Severity::Error,
    path: "slophammer.yml",
    message: "Configured Rust scope must cover all production files or exclude them with reasons",
    description: "Every production Rust file must be in configured scope or covered by a conventional or reasoned exclude, so findings cannot be hidden by narrowing scope.",
    tool: None,
    status: "implemented",
};
static RUST_SUPPRESSIONS_JUSTIFIED: RuleDefinition = RuleDefinition {
    id: rule_ids::RUST_SUPPRESSIONS_JUSTIFIED,
    title: "Rust suppressions justified",
    category: "rust",
    severity: Severity::Error,
    path: ".",
    message: "allow attributes in production Rust code must carry a reason",
    description: "An #[allow(...)] attribute in production scope must carry an adjacent // reason comment or use #[expect(..., reason = \"...\")]; bare suppressions accumulate silently.",
    tool: None,
    status: "implemented",
};
